fix: prevent AttributeError in AnalyzerAgent.run by storing agent ref…

soroushfathi opened the pull request:
### Technical Description

The `AnalyzerAgent.run` method previously appended raw coroutines to the `tasks` list.
Later, when iterating over results, the code attempted to access `tasks[i].agent.name`. Since `tasks[i]` was just a coroutine, this caused:

```
AttributeError: 'coroutine' object has no attribute 'agent'
```

#### Fix

* Updated `tasks` to store tuples of `(agent, coroutine)` instead of only the coroutine.
* Adjusted the `asyncio.gather` call to run only the coroutine part.
* Logging now retrieves the agent reference from the tuple, ensuring proper attribution of success/failure messages.

#### Benefits

* Prevents runtime `AttributeError` in logging.
* Makes the code more explicit by always pairing an agent with its coroutine task.
* Improves reliability when multiple analyzer agents run concurrently.

---

Do you want me to also suggest a **"Test Plan"** section you can add to show maintainers how you verified the fix works?

soroushfathi committed:
fix: prevent AttributeError in AnalyzerAgent.run by storing agent reference

soroushfathi committed:
refactor logging

miladosos commented:
Thanks for identifying and fixing this bug! The tuple approach works and correctly logs agent names. After reviewing our internal codebase, I wanted to share some alternative approaches that might be cleaner.

## About the Current Fix

Your tuple approach `(agent, coroutine)` correctly solves the AttributeError and maintains proper agent name logging. It works well! The main consideration is whether the tuple structure with index-based unpacking (`tasks[i][0]`, `t[1]`) is the most readable approach for this use case.

## Alternative Approaches

### Option 1: Simple Index Logging (Lowest Complexity)

Just log `Agent {i}` instead of agent names - this removes the `AttributeError` entirely without structural changes:

```python
for i, result in enumerate(results):
    if isinstance(result, Exception):
        Logger.error(f"Agent {i} failed: {result}")
    else:
        Logger.info(f"Agent {i} completed successfully")
```

This is what our internal version uses. Simple and effective.

### Option 2: Dictionary-Based Approach (Best for Debugging)

Use a dictionary with agent names as keys to maintain agent-to-task mapping:

```python
# Build dict of agent_name -> coroutine
agent_tasks = {}

if not self._config.exclude_structure:
    analysis_files.append(...)
    agent = self._structure_analyzer_agent
    agent_tasks[agent.name] = self._run_agent(
        agent=agent,
        user_prompt=self._render_prompt("agents.structure_analyzer.user_prompt"),
        file_path=self._config.repo_path / ".ai" / "docs" / "structure_analysis.md",
    )

if not self._config.exclude_dependencies:
    analysis_files.append(...)
    agent = self._dependency_analyzer_agent
    agent_tasks[agent.name] = self._run_agent(
        agent=agent,
        user_prompt=self._render_prompt("agents.dependency_analyzer.user_prompt"),
        file_path=self._config.repo_path / ".ai" / "docs" / "dependency_analysis.md",
    )

# ... repeat for other agents

Logger.debug("Running all agents")

# Run all agents concurrently
results = await asyncio.gather(*agent_tasks.values(), return_exceptions=True)

Logger.debug("All agents finished")

# Log results with meaningful agent names
for agent_name, result in zip(agent_tasks.keys(), results):
    if isinstance(result, Exception):
        Logger.error(f"Agent {agent_name} failed: {result}", exc_info=True)
    else:
        Logger.info(f"Agent {agent_name} completed successfully")
```

**Benefits over tuples:**
- Uses agent's existing `name` property (single source of truth)
- Dict keys are readable strings, easier to debug
- No need for index-based access or `.name` lookups
- More explicit and self-documenting than tuple unpacking

## My Recommendation

Your current fix works correctly! However, I'd prefer if you could refactor to **Option 2** (dictionary-based approach) for better readability and maintainability. The dict approach with agent names as keys is more explicit and easier to debug than tuple unpacking.

Would you be willing to update your PR to implement Option 2? Let me know if you have any questions or need help with the implementation!

soroushfathi committed:
refactor(analyzer): use dictionary-based agent task mapping for clearer logging

Replaced the tuple-based (agent, coroutine) approach with a dictionary
that maps agent names to their async tasks. This improves readability,
maintains explicit agent naming in logs, and removes the need for index-
based access when iterating over results.

Verified that all analyzer agents execute successfully and logs show
correct agent names.

soroushfathi commented:
Thanks for the detailed review and for sharing your internal approaches!
I’ve updated the PR to use the dictionary-based mapping (agent_name -> coroutine) as suggested.

This version keeps the implementation cleaner and more readable, while preserving correct logging and concurrent execution.
I also verified that all agents execute successfully and logs correctly include the agent names.

Let me know if you’d like any additional adjustments or style tweaks!

soroushfathi committed:
Merge branch 'main' into fix/analyzer-run-logging
